Rakefile

```diff
@@ -1,66 +1,112 @@
-require "json"
 require "pathname"
+require "time"
 
-PROJ_DIR     = Pathname.new(__dir__)
-CONFIG_FILE  = PROJ_DIR.join(ENV["CNFFILE"] || "config.json")
-BUILD_ROOT   = Pathname.new(ENV["OBJDIR"] || "build")
-INSTALL_ROOT = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+PROJ_DIR      = Pathname.new(__dir__)
+APPS_DIR      = PROJ_DIR.join('apps')
+GEMFILE       = PROJ_DIR.join('Gemfile')
+INSTALL_ROOT  = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+VENDOR_BUNDLE = (ENV['VENDOR_BUNDLE'] == "yes" || ENV['VENDOR_BUNDLE'] == "true")
 
-def all_components
-  JSON.parse(CONFIG_FILE.read).map { |c| Component.new(c) }
+def apps
+  Dir["#{APPS_DIR}/*"].map { |d| Component.new(d) }
 end
 
 class Component
   attr_reader :name
-  attr_reader :url
-  attr_reader :tag
-
-  def initialize(opts = {})
-    @name = opts.fetch("name") { raise ArgumentError, "No name specified. Missing argument: name" }.to_s
-    @url  = opts.fetch("url")  { raise ArgumentError, "No url specified. Missing argument: url" }.to_s
-    @tag  = opts.fetch("tag")  { raise ArgumentError, "No tag specified. Missing argument: tag" }.to_s
-    @app  = opts.fetch("app", nil)
+  attr_reader :path
+
+  def initialize(app)
+    @name = File.basename(app)
+    @path = Pathname.new(app)
   end
 
-  def app?
-    !!@app
+  def ruby_app?
+    @path.join('config.ru').exist?
   end
 
-  def build_root
-    app? ? BUILD_ROOT.join("apps", name) : BUILD_ROOT.join(name)
+  def node_app?
+    @path.join('app.js').exist?
   end
+end
 
-  def download_url
-    "#{url}/archive/#{tag}.tar.gz"
+desc "Package OnDemand"
+task :package do
+  `which gtar 1>/dev/null 2>&1`
+  if $?.success?
+    tar = 'gtar'
+  else
+    tar = 'tar'
   end
+  version = ENV['VERSION']
+  if ! version
+    latest_commit = `git rev-list --tags --max-count=1`.strip[0..6]
+    latest_tag = `git describe --tags #{latest_commit}`.strip[1..-1]
+    datetime = Time.now.strftime("%Y%m%d-%H%M")
+    version = "#{latest_tag}-#{datetime}-#{latest_commit}"
+  end
+  sh "git ls-files | #{tar} -c --transform 's,^,ondemand-#{version}/,' -T - | gzip > packaging/v#{version}.tar.gz"
 end
 
-task :default => :build
+namespace :build do
+  desc "Build gems"
+  task :gems do
+    if VENDOR_BUNDLE
+      args = "--path vendor/bundle"
+    else
+      args = ""
+    end
+    apps.each do |a|
+      next unless a.ruby_app?
+      chdir a.path do
+        sh "bundle install #{args}"
```

Comment:
I think its preferable to use`bin/bundle install` instead of `bundle install` but I do not remember the ramifications of this

Reply:
I don't think it's very different since bin/bundle was just this:

```
#!/usr/bin/env ruby
ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../../Gemfile', __FILE__)
load Gem.bin_path('bundler', 'bundle')
```

The only feature lost is BUNDLE_GEMFILE environment variable but doing that with `rake build` to change path would result in all Ruby apps being built with same Gemfile unless you did `BUNDLE_GEMFILE=... rake build:myjobs`. Should I put `bin/bundle` back for each app?

Reply:
I remember occasions where building a single app on webtest02 failed when using `bundle install` and succeeded when using `bin/bundle install`. It is possible this type of problem will not be observed when installing gems to GEM_HOME. It might have occurred when some form of bundler or other gem accidentally got installed to the developer's home directory, so then `bin/bundle` avoided the conflict. We can leave it as it is though and if we run into problems change it to `bin/bundle install`.